### PR TITLE
Remove reflective required field checker in favor of code-gen.

### PR DESCRIPTION
**Pull request by JakeWharton**
This removes reflection from the message building path and keeps it only for serialization. In addition to not using reflection for this check, we eliminate the Map caching the builder adapter instances and all the state each instance kept.

### CODE REVIEW

File: wire-runtime/src/test/java/com/squareup/wire/protobuf/TestAllTypes.java [hunk 1]
```diff
@@ -517,7 +518,7 @@ public void testRequiredFields() {
       builder.build();
       fail();
     } catch (IllegalStateException e) {
-      assertEquals("Required fields not set:\n  req_bool\n  req_int32\n  req_sfixed64",
+      assertEquals("Required fields not set:\n  req_int32\n  req_sfixed64\n  req_bool",
```
Comment (JakeWharton) [inline wire-runtime/src/test/java/com/squareup/wire/protobuf/TestAllTypes.java]: This is now in proto declaration order

**Comment by danrice-square** [comment]
Love it!

File: wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java [hunk 2]
```diff
@@ -661,9 +665,21 @@ private MethodSpec builderBuild(WireMessage message, ClassName javaType) {
...
         .returns(javaType);
-    if (message.hasRequiredFields()) {
-      result.addStatement("checkRequiredFields()");
+
+    List<WireField> requiredFields = message.getRequiredFields();
+    if (!requiredFields.isEmpty()) {
+      List<String> conditionals = new ArrayList<String>();
+      List<String> missingArgs = new ArrayList<String>();
+      for (WireField requiredField : requiredFields) {
+        conditionals.add(requiredField.name() + " == null");
```
Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java]: Consider CodeBlock.Builder!

File: wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java [hunk 3]
```diff
@@ -661,9 +665,21 @@ private MethodSpec builderBuild(WireMessage message, ClassName javaType) {
...
+        missingArgs.add(requiredField.name() + ", \"" + requiredField.name() + '"');
```
Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java]: Then you could use `$S`, which will Do The Right Thing if requiredField.name() needs any escaping.

File: wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java [hunk 4]
```diff
@@ -649,7 +651,9 @@ private MethodSpec builderSetExtension(ClassName javaType, ClassName builderType
   //
   // @Override
   // public SimpleMessage build() {
-  //   checkRequiredFields();
+  //   if (field_one == null || field_two == null) {
```
Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java]: nit: Example is missing `\n`

File: wire-compiler/src/main/java/com/squareup/wire/model/WireMessage.java [hunk 5]
```diff
@@ -61,11 +63,14 @@
     return fields;
   }
 
-  public boolean hasRequiredFields() {
+  public List<WireField> getRequiredFields() {
+    List<WireField> required = new ArrayList<WireField>();
```
Comment (swankjesse) [inline wire-compiler/src/main/java/com/squareup/wire/model/WireMessage.java]: ImmutableLists!

File: wire-runtime/src/main/java/com/squareup/wire/Message.java [hunk 6]
```diff
@@ -298,10 +298,24 @@ UnknownFieldMap ensureUnknownFieldMap() {
...
+     * Create an exception for missing required fields.
+     *
+     * @param args Alternating field value and field name pairs.
      */
-    public void checkRequiredFields() {
-      WIRE.builderAdapter(getClass()).checkRequiredFields(this);
+    protected IllegalStateException missingRequiredFields(Object... args) {
+      StringBuilder sb = new StringBuilder();
+      String plural = "";
+      // Avoid creating an iterator
```
Comment (swankjesse) [inline wire-runtime/src/main/java/com/squareup/wire/Message.java]: The comment really isn't necessary

File: wire-runtime/src/test/java/com/squareup/wire/protobuf/TestAllTypes.java (hunk 1 above)
Comment (swankjesse) [inline wire-runtime/src/test/java/com/squareup/wire/protobuf/TestAllTypes.java]: Nice.

**Comment by swankjesse** [comment]
LGTM

File: wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java (hunk 4 above)
Comment (JakeWharton) [inline wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java]: I was hiding from the shame of the two-field case that looks so ugly :sob:

File: wire-runtime/src/main/java/com/squareup/wire/Message.java (hunk 6 above)
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/Message.java]: And it's not even true anymore. Came along in the copy/paste party.

File: wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java (hunk 2 above)
Comment (JakeWharton) [inline wire-compiler/src/main/java/com/squareup/wire/java/TypeWriter.java]: Done. So much nicer!

**Comment by JakeWharton** [comment]
Travis took the day off.
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Wire Protocol Buffers .............................. SUCCESS [  3.744 s]
[INFO] Wire Protocol Buffer Runtime ....................... SUCCESS [  6.891 s]
[INFO] Wire Protocol Buffer Compiler ...................... SUCCESS [  5.368 s]
[INFO] Wire Protocol Buffer Gson Support .................. SUCCESS [  1.248 s]
[INFO] Wire Protocol Buffer Maven Plugin .................. SUCCESS [  6.673 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 24.103 s
[INFO] Finished at: 2015-07-06T18:21:22-04:00
[INFO] Final Memory: 39M/485M
[INFO] ------------------------------------------------------------------------
```

**Pull request merged**